### PR TITLE
Add ListContext to ReferenceArrayInput

### PR DESCRIPTION
**Pull request by djhi**
- [x] Implementation
- [x] Ability to select rows to change the input value
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
This should be PRed against `next`

File: packages/ra-core/src/controller/input/useReferenceArrayInputContext.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+export const useReferenceArrayInputContext = <
+    T extends ReferenceArrayInputContextValue = ReferenceArrayInputContextValue
+>(
+    props: T
+): ReferenceArrayInputContextValue => {
+    const context = useContext(ReferenceArrayInputContext);
+
+    if (props.choices) {
+        return props;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceArrayInputContext.ts:18]: Please use a more fine-grained approach, as in useListContext in master

File: packages/ra-core/src/util/indexById.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { Record, RecordMap } from '../types';
+
+export const indexById = (records: Record[] = []): RecordMap =>
```
Comment (fzaninotto) [inline packages/ra-core/src/util/indexById.ts:3]: please add jsDoc with sample input/output

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx [hunk 3]
```diff
@@ -141,19 +143,33 @@ const ReferenceArrayInput: FC<ReferenceArrayInputProps> = ({
         input,
     });
 
+    const listContext = useMemo(
+        () => ({
+            ...controllerProps,
+            setSort: controllerProps.setSortForList,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx:149]: Please add a jsDoc explaining why you do some voodoo here

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx [hunk 4]
```diff
@@ -141,19 +143,33 @@ const ReferenceArrayInput: FC<ReferenceArrayInputProps> = ({
...
-            isRequired={isRequired}
-            meta={meta}
-            translate={translate}
-            children={children}
-            {...props}
-            {...controllerProps}
-        />
+        <ResourceContextProvider value={props.reference}>
+            <ReferenceArrayInputContextProvider value={controllerProps}>
+                <ListContextProvider value={listContext}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx:159]: I don't get it. What's the use of a ReferenceArrayInputContext if all you need is already in the ListContext?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx (hunk 4 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx:159]: I agree it is mostly redundant now but it does provide the `warning` prop which indicates non "fatal" errors regarding references.

File: packages/ra-core/src/util/indexById.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/util/indexById.ts:17]: Done

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx (hunk 3 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx:152]: Done

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**